tests/test_cache.py

```diff
@@ -0,0 +1,215 @@
+import httpretty
+import json
+import os
+import pickle
+import redis
+import requests
+import shutil
+import struct
+import time
+import unittest
+from cachecontrol.caches.file_cache import FileCache
+from cachecontrol.caches.redis_cache import RedisCache
+from canonicalwebteam.http import CachedSession
+from datetime import timedelta
+from mockredis import mock_redis_client
+from unittest.mock import patch
+
+
+file_cache_directory = ".testcache"
+
+
+class MockRedisSingleton:
+    def __init__(self):
+        self.redis_client = None
+
+    def mock_redis_client_singleton(self, *args, **kwargs):
+        if not self.redis_client:
+            self.redis_client = mock_redis_client()
+        return self.redis_client
+
+
+class TestCachedSession(unittest.TestCase):
+    mock_redis_singleton = MockRedisSingleton()
+
+    def tearDown(self):
+        try:
+            shutil.rmtree(file_cache_directory)
+        except:
+            pass
+
+    @httpretty.activate
+    def test_custom_heuristic(self):
+        def request_callback(request, uri, response_headers):
+            return [200, response_headers, json.dumps({"epoch": time.time()})]
+
+        httpretty.register_uri(
+            httpretty.GET, "https://now.httpbin.org", body=request_callback
+        )
+
+        session = CachedSession(
+            fallback_cache_duration=2,
+            file_cache_directory=file_cache_directory,
+        )
+
+        # with a 2s retention, and a 1.1s time between requests, 2 of the
```

Comment:
We do want our tests to run as quickly as possible. Have you tried with even smaller values? If it allows you, it would be good to test with fractions of a second, since I imagine the execution time between calls will only be nanoseconds. Would it let you, for instance, make the `fallback_cache_duration=0.2`? And then sleep for `0.1`?

Reply:
I have tried it, but unfortunately the cache did not seem to work. 

[The Mozilla docs](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Cache-Control#Expiration) say that it is given in seconds, so sub-second values might not work correctly. My best guess here.

Reply:
It seems that the tests still perform inconsistently.
I can not figure it out at the moment, a second pair of eyes on this would be good. Maybe a pairing session.
The false negatives could be really annoying further down the line.

Reply:
Fixed by using [freezegun](https://github.com/spulec/freezegun)